Add roof orientation quest

FloEdelmann opened the pull request:
Specifying the roof orientation seems to be an often requested feature for StreetComplete:

- https://github.com/streetcomplete/StreetComplete/issues/1063
- https://github.com/streetcomplete/StreetComplete/issues/3164
- https://github.com/streetcomplete/StreetComplete/issues/4197
- https://github.com/streetcomplete/StreetComplete/issues/5178

It was actually not rejected as a quest for StreetComplete, but for good usability, a preview of the selected roof orientation was deemed necessary: https://github.com/streetcomplete/StreetComplete/issues/1063#issuecomment-882070270

While I agree with that statement for StreetComplete, IMO it's not that important in SCEE and also mitigated when limiting the quest to `roof:shape = gabled`. Since no progress has been made towards that preview, I think it's a good idea to add a basic quest (without a preview) to SCEE:

<img src="https://github.com/user-attachments/assets/0078f55c-5d5d-416f-8255-d6ed825428d1" width="300">

---

By the way: Implementing this quest was inspired by https://community.openstreetmap.org/t/streets-gl-a-new-3d-renderer-for-osm/98594/192 and (again) looking through the 3D rendered houses in my neighborhood, where many terraced houses have the wrong default orientation in Streets GL.

FloEdelmann committed:
Add roof orientation quest

FloEdelmann committed:
Rename and adjust `default_disabled_msg_roofColour` string

FloEdelmann committed:
Translate strings to German

mcliquid commented on app/src/main/res/values-de/strings_ee.xml:
```diff
@@ -632,4 +632,7 @@
     <string name="pref_custom_geometry_title">Zeige Map-Features aus GeoJson-Datei an</string>
     <string name="pref_custom_geometry_info">Wenn du Text anzeigen möchtest, muss er in der Eigenschaft \"name\" enthalten sein</string>
     <string name="pref_tile_maxzoom">Maximal unterstützter Zoom</string>
+    <string name="quest_roofOrientation_along">Parallel zur längeren Gebäudeseite</string>
```
Aren't these strings overwritten by Weblate?

FloEdelmann replied:
According to https://github.com/Helium314/SCEE/pull/669#discussion_r1927406768, adding them right away should work.

mcliquid replied:
It works, but you can't adjust the translation in Weblate :(

FloEdelmann replied:
So the strings added via GitHub don't appear for translation in Weblate at all? Or they are just empty in Weblate and have to be translated from scratch, and while they are not translated there, the translations added in GitHub take precedence?

Helium314 replied:
> It works, but you can't adjust the translation in Weblate :(

Oh, I wasn't aware of that. I remember disabling the possibility of translating the original strings, maybe this affects everything that was not added in Weblate?

mcliquid replied:
We can just try it out with this PR, what exactly is happening in Weblate, then we will know for the future :)

Helium314 replied:
You can report back what happens.
Actually I only saw later that this PR also _changes_ translations. This is a bit trickier, as it may create merge conflicts with Weblate. Pulling translations right before merging shoulid be fine, but even like that I remember having issues in one case.

FloEdelmann replied:
> this PR also _changes_ translations

Does it though? Since the string resource is also renamed, it's more like _deleting_ an existing string (with all translations) and _adding_ a new string (with most translations).

Helium314 commented:
I can already smell the issues about (nearly) square buildings, or buildings with weird shapes... but trying to excluce such cases is not worth the relateive large effort.

FloEdelmann commented:
I tried it today outside, and I agree it's often hard to answer for non-rectangle or nearly-square buildings. Adding logic to exclude those is likely not worth the effort, but I'll give it a try anyway.

FloEdelmann committed:
Only ask for nearly rectangular but not nearly square roofs

FloEdelmann commented:
It was a bit of a challenge, but I am quite happy with the result :slightly_smiling_face:
See the [test](https://github.com/Helium314/SCEE/pull/733/files#diff-5193e843885c6faf5aef148fc5770a2dae4f29d6263b465a6ee08cb089dd1cf0) I added. It verifies that the quest is now …

* asked for rectangular roofs like https://osm.org/way/128296730
* asked for nearly rectangular roofs like https://osm.org/way/525058074 or https://osm.org/way/95462050
* not asked for nearly square roofs like https://osm.org/way/93389455
* not asked for other odd-shaped roofs like https://osm.org/way/790940999 or https://osm.org/way/128115590

Helium314 commented:
Wow, that certainly doesn't look like low effort. But seems to nicely find nearly rectangular buildings with more than 4 nodes!

Helium314 commented:
Though maybe some optimization would be useful?
```
17:52:53.596 OsmQuestController       D  ShowFixme: Found 10 quests in 37ms
17:52:53.602 OsmQuestController       D  ShowTrafficStuff: Found 647 quests in 110ms
17:52:53.618 OsmQuestController       D  ShowSeating: Found 181 quests in 70ms
17:52:53.630 OsmQuestController       D  ShowOther: Found 206 quests in 125ms
17:52:54.023 OsmQuestController       D  CheckShopExistence: Found 101 quests in 1398ms
17:52:58.263 OsmQuestController       D  AddRoofOrientation: Found 13 quests in 5269ms
```
I only found 2 quests where it took more than 1s, so AddRoofOrientation stands out quite clearly.

FloEdelmann committed:
Return early

FloEdelmann committed:
More idiomatic Kotlin

FloEdelmann committed:
Fix circumference calculation

FloEdelmann committed:
Avoid distance recalculation

FloEdelmann committed:
Reduce number of loop iterations

FloEdelmann committed:
Avoid distance recalculation again

FloEdelmann committed:
Remove `QuadrangleSides` data class

FloEdelmann committed:
Avoid max recalculation

FloEdelmann commented:
I think I could reduce the time by around 25%, mostly by using early returns. However my test times were somewhat noisy, so could you please check again @Helium314?

FloEdelmann committed:
Add quest to chonkerIndex

Helium314 commented:
I did not yet check, but I noticed that performance might profit from using `flatDistanceTo(Arc)` instead of `distanceTo(Arc)`. The error from assuming a locally flat earth should be negligible, unless you're mapping _huge_ buildings or very close to north/southpole.

FloEdelmann committed:
Act like the earth is flat because it makes computation much faster

and for such small distances it's a fine approximation

Helium314 commented:
With the improvements, I now get around 2300 ms.
Digging into where it takes long revealed that buildings with many nodes require disproportionately longer, while being unlikely to get the roof shape quest.

Limitiing the number of nodes to 20 reduces quest creation time to 300 ms, and 200 ms at 15, while keeping the number of applicable buildings at 19.
Using 10 already kicks auf 6 of the 19 buildings in my test area, so that would definitely be too strict.

2 other areas:
* mostly small and simple buildings
  * max 10 nodes: 66 quests in 40 ms
  * max 15 nodes: 70 quests in 90 ms
  * max 20 nodes: 71 quests 200 ms
  * no limit: 71 quests in 200 ms
* mostly medium size somewhat more complex buildings, few very complex buildings
  * max 15 nodes: 12 quests in 25 ms
  * max 20 nodes: 13 quests 50 ms
  * no limit: 15 quests in 1800 ms

So limiting the number of nodes to somewhere in the range of 15-20 seems like a reasonable shortcut to me.

FloEdelmann committed:
Exclude complex buildings

FloEdelmann committed:
Revert "Add quest to chonkerIndex"

This reverts commit 85cd4de884d5927a8df71f6ffdcd1a61fe4a0de4.

FloEdelmann commented:
Thanks for checking @Helium314! I added that check and updated the tests accordingly. This should now be ready to be merged :slightly_smiling_face: 

However, before doing that, I wonder if it would now actually be suitable for upstream StreetComplete. @westnordost what do you think?

To quickly summarize this PR: It adds a new roof orientation quest (seems to be highly requested, see links in PR description) that is only asked for `roof:shape = gabled` for buildings that are roughly rectangular and not square, so it should always be easy to answer. Thanks to @Helium314's suggestions, we could reduce the search time down to a reasonable level.

HolgerJeromin commented on app/src/test/java/de/westnordost/streetcomplete/quests/roof_orientation/AddRoofOrientationTest.kt:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    @Test fun `not applicable to way with 3 points`() {
+        // https://osm.org/way/926313667
```
Do we want to link to the exact version?
https://osm.org/way/926313667/history/2
Perhaps the way is reused/changed.

FloEdelmann replied:
I thought about that, too, but

1. it is just for quick visualization, so it's not that bad if it breaks
2. linking to an exact version doesn't highlight the building outline on the map
3. that is the history of the way; does it even change if only one of its member nodes changes position? If not, then linking to an exact version would be confusing.

HolgerJeromin replied:
It would be different if the history url would show the old geometry on the map.
But fine with keeping it this way right now.

FloEdelmann committed:
Merge branch 'modified' into roof-orientation

FloEdelmann commented:
@westnordost sorry for bothering, but maybe you missed my comment above?

westnordost commented:
Well, my original vision was that the roof shape plus roof orientation (or maybe even the building levels) quests could be merged into one, featuring a powerful preview in 3D.

[I have been in discussion with](https://github.com/tordanik/OSM2World/issues/171) @tordanik, the author of OSM2World, whether it would be possible (one day) to have that project as a library that given some geometry and tags, it would return a renderable 3D model. Now, recently, he got a grant by the prototype fund so it had become ever more likely that something like that would be possible, but alas, we are moving away from Java and Osm2World is Java, so this hardly seems possible anymore.

Probably, the most likely thing that could happen here is to actually implement it oneself, the most complex part being to calculate the 3D roof shape, of course, basically (just) copy&pasting the relevant parts from OSM2World and converting these to Kotlin. I imagine that it is not *that* much effort, but probably still much more than anyone would be ready to invest for just one quest in StreetComplete.

Anyway.

I see the following (solvable, with different effort) issues with a roof shape orientation quest:

- 💤 **spammy**: most roof ridges will be alongside the longest side. For this kind of thing, it would be more helpful if one could see an (editable) map with how the buildings are rendered. Not really possible in StreetComplete, though. This issue is mitigated a bit by this being disabled by default.

- 🐿️ **difficult**: first, the question and answer format without any kind of preview is difficult, as to describe it precisely, words are used that are basically expert vocabulary. Second, the user, standing in front of the building, doesn't know which is the longest and which is the shortest side of the building. **Only** by looking on a kind of preview or on the map, he'll know. Which of course begs the question whether this is a good fit for StreetComplete at all, if one _needs_ satellite imagery / a map anyway. (see **💵 Valuable Surveyors**)

Especially the second point can be mitigated. You added a lot of code to exclude those buildings where it is not clear on the map which is the shortest and which is the longest side of the building. Also, to exclude those that are not really rectangular but more complex.

It would be ideal to instead of having complicated wording and filtering exactly those buildings (squarish buildings) for which the information would make most sense, have a preview. Show, don't tell.
This doesn't have to be a 3D preview, not at all. It could be a very simple custom `Drawable` that shades the roof in two different colors

![imagen](https://github.com/user-attachments/assets/fa3a47c5-d123-45d1-b5b9-0b398c29daf3)

on either side of a line drawn from the midpoints of the short/long side through the center point. For relatively simple rectangles, this visualization would do just fine. The resulting `RoofDrawable` would be much less complex than e.g. the `StreetParkingDrawable` and even the latter is not **that** long in terms of lines of code.

IMO it would be definitely worth it to rather show two such `RoofDrawables` instead of trying to convey it in text.

The algorithm that filters out too complex buildings could then filter out those buildings where either there are no opposing long sides/short sides or where a line drawn from the midpoint to the other side would intersect (or come close to?) the building outline itself (banana shaped building). Or, be more strict, if it results in less code. Most buildings will be a rectangle anyway.

---

tl;dr: at the current state, I would evaluate this quest to be problematic and a source of future issues and calls to mitigate it. So while I think this quest could in general go into StreetComplete, I would like to forestall this and have these issues already solved at the very start.

Helium314 committed:
Merge branch 'modified' into roof-orientation
